Sync `v2-2-stable` with `v2-2-test` to release `2.2.5`

ephraimbuddy opened the pull request.

bbovenzi committed:
Fix: Do not render undefined graph edges (#19684)

* Fix: Do not render undefined graph edges

A user had an issue where a `targetId` was undefined and that caused the whole graph view to crash. Instead, we should check for the source and target before rendering the edge.

* move all checks to one line

(cherry picked from commit bd109b4336d5950168b6ff04a752290e437b2b8c)

ashb committed:
Update Kubernetes library version (#18797)

Previously we pinned this version as v12 as a change to Kube library
internals meant v1.Pod objects now have a logger object inside them, and
couldn't be pickled on Python 3.6.

To fix that we have "backported" the change in Python 3.7 to make Logger
objects be pickled "by name". (In Python 3.7 the change adds
`__reduce__` methods on to the Logger and RootLogger objects, but here
we achieve it `copyreg` stdlib module so we don't monkeypatch
anything.)

This fix is also applied in to airflow core in a separate commit, but we
also apply it here in the provider so that cncf.kubernetes client
library can be updated but still used with older versions of Airflow
that don't have this fix in.

(cherry picked from commit 7222f68d374787f95acc7110a1165bd21e7722a1)

dstandish committed:
Rename `to_delete` to `to_cancel` in TriggerRunner (#20658)

The queue's purpose is to track triggers that need to be canceled. The language `to_delete` was a bit confusing because for one it does not actually delete them but cancel them.  The deletion work is actually in `cleanup_finished_triggers`.  It seems that this method will usually not do anything and it's only for cancelling triggers that are currently running but for whatever reason no longer should be.  E.g. when a task is killed and therefore the trigger is no longer needed, or some multi-triggerer scenarios.  So putting cancel in the name also highlights that this is about stopping running triggers, not e.g. purging completed ones.

(cherry picked from commit c20ad79b40ea2b213f6dca221221c6dbd55bd08f)

dstandish committed:
Fix duplicate trigger creation race condition (#20699)

The process for queueing up a trigger, for execution by the TriggerRunner, is handled by the TriggerJob's `load_triggers` method.  It fetches the triggers that should be running according to the database, checks if they are running and if not it adds them to `TriggerRunner.to_create`.  The problem is tha there's a small window of time between the moment a trigger (upon termination) is purged from the `TriggerRunner.triggers` set,  and the time that the database is updated to reflect the trigger's doneness.  If `TriggerJob.load_triggers` runs during this window, the trigger will be added back to the `TriggerRunner.to_create` set and it will run again.

To resolve this what we do here is, before adding a trigger to the `to_create` queue, instead of comparing against the "running" triggers, we compare against all triggers known to the TriggerRunner instance.  When triggers move out of the `triggers` set they move into other data structures such as `events` and `failed_triggers` and `to_cancel`.  So we union all of these and only create those triggers which the database indicates should exist _and_ which are know already being handled (whatever state they may be in) by the TriggerRunner instance.

(cherry picked from commit 16b8c476518ed76e3689966ec4b0b788be935410)

subkanthi committed:
Set X-Frame-Options header to DENY only if X_FRAME_ENABLED is set to true. (#19491)

(cherry picked from commit 084079f446570ba43114857ea1a54df896201419)

mike-mcdonald committed:
fix: Update custom connection field processing (#20883)

* fix: Update custom connection field processing

Fixes issue where custom connectionfields are not updated because `extra` field is in form and has previous values, overriding custom field values.
Adds portion of connection form tests to test functionality.

(cherry picked from commit 44df1420582b358594c8d7344865811cff02956c)

uranusjr committed:
Ensure clear_task_instances sets valid run state (#21116)

(cherry picked from commit d17db3ce8ee8a8a724ced9502c73bc308740a358)

malthe committed:
A trigger might use a connection; make sure we mask passwords (#21207)

(cherry picked from commit 3d0c1aea5a85a4d31d3ade530e4c5b85b045503a)

dstandish committed:
Fix trigger dag redirect from task instance log view (#21239)

(cherry picked from commit fd8f21f1a3c1e9f447a96a193adf5d28e207731a)

josh-fell committed:
Update `ExternalTaskSensorLink` to handle templated `external_dag_id` (#21192)

(cherry picked from commit 8da7af2bc0f27e6d926071439900ddb27f3ae6c1)

pingzh committed:
Fix Resources __eq__ check (#21442)

(cherry picked from commit 6b308446eae2f83bf379f976c7d7801aa53370a3)

pingzh committed:
Dispose unused connection pool (#21565)

(cherry picked from commit 8155e8ac0abcaf3bb02b164fd7552e20fa702260)

malthe committed:
Fix race condition between triggerer and scheduler (#21316)

(cherry picked from commit 2a6792d94d153c6f2dd116843a43ee63cd296c8d)

KevinYanesG committed:
Fix postgres part of pipeline example of tutorial (#21586)

(cherry picked from commit 40028f3ea3e78a9cf0db9de6b16fa67fa730dd7a)

uranusjr committed:
Log exception in local executor (#21667)

(cherry picked from commit a0fb0bbad312df06dd0a85453bd4f93ee2e01cbb)

mnojek committed:
Extend documentation for states of DAGs & tasks and update trigger rules docs (#21382)

(cherry picked from commit 4e959358ac4ef9554ff5d82cdc85ab7dc142a639)

bbovenzi committed:
Fix graph autorefresh on page load (#21736)

* fix auto refresh check on page load

* minor code cleanup

* remove new line

(cherry picked from commit b2c0a921c155e82d1140029e6495594061945025)

uranusjr committed:
Fix stray order_by(TaskInstance.execution_date) (#21705)

(cherry picked from commit bb577a98494369b22ae252ac8d23fb8e95508a1c)

msumit committed:
Fix triggerer --capacity parameter (#21753)

(cherry picked from commit 9076b67c05cdba23e8fa51ebe5ad7f7d53e1c2ba)

norm committed:
Fix the triggerer capacity test (#21760)

Commit 9076b67 changed the triggerer logic to use int not string.

(cherry picked from commit e1fe30c70d0fe9c033db9daf9d4420f7fa815b2d)

zmac12 committed:
Correct a couple grammatical errors in docs (#21750)

Just reading through the docs as we implement Airflow on our end, saw a couple additions that could be made.

(cherry picked from commit 3bb63d4cfbbb534298d32ec987f25a02c11fc4e6)

ephraimbuddy committed:
Fix max_active_runs=1 not scheduling runs when min_file_process_interval is high (#21413)

The finished dagrun was still being seen as running when we call dag.get_num_active_runs
because the session was not flushed. This PR fixes it

(cherry picked from commit feea143af9b1db3b1f8cd8d29677f0b2b2ab757a)

kadai0308 committed:
Correctly handle multiple '=' in LocalFileSystem secrets. (#21694)

(cherry picked from commit 919b75ba20083cc83c4e84e35aae8102af2b5871)

Commit:
extends typing-extensions to be installed with python 3.8+ #21566 (#21567)

(cherry picked from commit e4ead2b10dccdbe446f137f5624255aa2ff2a99a)

jkramer-ginkgo committed:
Fix assignment of unassigned triggers (#21770)

Previously, the query returned no alive triggerers which resulted
in all triggers to be assigned to the current triggerer. This works
fine, despite the logic bug, in the case where there's a single
triggerer. But with multiple triggerers, concurrent iterations of
the TriggerJob loop would bounce trigger ownership to whichever
loop ran last.

Addresses https://github.com/apache/airflow/issues/21616

(cherry picked from commit b26d4d8a290ce0104992ba28850113490c1ca445)

malthe committed:
Truncate stack trace to DAG user code for exceptions raised during execution (#20731)

(cherry picked from commit 7ea0f7694f39358583cf8f0f8af896fc776bdffe)

hubert-pietron committed:
Fix logging JDBC SQL error when task fails (#21540)

(cherry picked from commit bc1b422e1ce3a5b170618a7a6589f8ae2fc33ad6)

malthe committed:
Log traceback in trigger excs (#21213)

(cherry picked from commit 4ad21f5f7c2d416cf813a860564bc2bf3e161d46)

mixilchenko committed:
Fix filesystem sensor for directories (#21729)

Fix walking through wildcarded directory in `FileSensor.poke` method

(cherry picked from commit 6b0ca646ec849af91fe8a10d3d5656cafa3ed4bd)

potiuk committed:
Fix handling some None parameters in kubernetes 23 libs. (#21905)

Kubernetes 23.* is more picky when it comes to values passed to
Pod Generator - it requires:

* imagePullPolicy
* dnsPolicy
* restartPolicy

to be not None.

We are fixing it in the way, that we simply skip setting those
if they are None.

(cherry picked from commit e856aba9ab69094787dfd0f6e911f20782069e92)

millin committed:
Fix incorrect data provided to tries & landing times charts (#21928)

(cherry picked from commit 2c57ad4ff9ddde8102c62f2e25c2a2e82cceb3e7)

potiuk committed:
Change the storage of frame to use threadLocal rather than Dict (#21993)

There is a very probable WeakKeyDict bug in Python standard
library (to be confirmed and investigated further) that
manifests itself in a very rare failure of the
test_stacktrace_on_failure_starts_with_task_execute_method

This turned out to be related to an unexpected behaviour
(and most likely a bug - to be confirmed) of WeakKeyDict
when you have potentially two different objects with the
same `equals` and `hash` values added to the same
WeakKeyDict as keys.

More info on similar report (but raised for a bit different
reason) bug in Python can be found here:

https://bugs.python.org/issue44140

We submitted a PR to fix the problem found
https://github.com/python/cpython/pull/31685

(cherry picked from commit 1949f5d76b5842d56db91c868ae4655bb7a7689f)

jedcunningham committed:
DB upgrade is required when updating Airflow (#22061)

Just strengthen the language that it is "required", not "recommended" to
run `airflow db upgrade` when upgrading Airflow versions.

(cherry picked from commit e7fed6bb3d7c8def86b47204176cebbfc6c401ff)

subkanthi committed:
Fixed dask executor and tests (#22027)

 Fixed dask executor and tests, distributed package does not ship with tests folder and the certificates, added certificates to certs folder

(cherry picked from commit d3c168c5a341327bb55eb855433620249941584a)

potiuk committed:
Add the new Airflow Trove Classifier to setup.cfg (#22241)

We have new Trove Classifiers in PyPI for Apache Airflow:

https://github.com/pypa/trove-classifiers/pull/87

This PR adds it for Airflow. The next release of Providers will
add the classifiers for providers.

(cherry picked from commit d3c65b64e9ff4c24a1a899af0d5eba954775f646)

xyu committed:
Filter out default configs when overrides exist. (#21539)

* Filter out default configs when overrides exist.

When sending configs to Airflow workers we materialize a temp config file. In #18772 a feature was added so that `_cmd` generated secrets are not written to the files in some cases instead favoring maintaining the raw `_cmd` settings. Unfortunately during materializing of the configs via `as_dict()` Airflow defaults are generated and materialized as well including defaults for the non `_cmd` versions of some settings. And due to Airflow setting precedence stating bare versions of settings winning over `_cmd` versions it results in `_cmd` settings being discarded:
https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html

This change checks `_cmd`, env, and secrets when materializing configs via `as_dict()` so that if the bare versions of the values is exactly the same as Airflow defaults and we have "hidden" / special versions of these configs that are trying to be set we remove the bare versions so that the correct version can be used.

Fixes: #20092
Related to: #18772 #4050

(cherry picked from commit e07bc63ec0e5b679c87de8e8d4cdff1cf4671146)

chenglongyan committed:
Disable default_pool delete on web ui (#21658)

(cherry picked from commit df6058c862a910a99fbb86858502d9d93fdbe1e5)

dstandish committed:
Remove RefreshConfiguration workaround for K8s token refreshing (#20759)

A workaround was added (https://github.com/apache/airflow/pull/5731) to handle the refreshing of EKS tokens.  It was necessary because of an upstream bug.  It has since been fixed (https://github.com/kubernetes-client/python-base/commit/70b78cd8488068c014b6d762a0c8d358273865b4) and released in v21.7.0 (https://github.com/kubernetes-client/python/blob/master/CHANGELOG.md#v2170).

(cherry picked from commit 7bd165fbe2cbbfa8208803ec352c5d16ca2bd3ec)

kaxil committed:
Add back celery intersphinx mapping (#22370)

We had disabled this previously in (#22254) but now the website is up on a different domain as listed in https://github.com/celery/celeryproject/issues/51#issuecomment-1072248499

(cherry picked from commit 1f7836e07e643b3e9e0eca57ae077072358c63c1)

davidhagens committed:
Fix broken links to celery documentation (#22364)

The celery documentation have been moved from https://docs.celeryproject.org/ to https://docs.celeryq.dev/. The old links now refer to a 404 error page, the new links to the actual documentation.

(cherry picked from commit a8de170c9fb31d219ec4287f79cfa58cc0b44866)

SamWheating committed:
Reduce DB load incurred by Stale DAG deactivation (#21399)

Deactivating stale DAGs periodically in bulk

By moving this logic into the DagFileProcessorManager and running it across all processed file periodically, we can prevent the use of un-indexed queries.

The basic logic is that we can look at the last processed time of a file (for a given processor) and compare that to the last_parsed_time of an entry in the dag table. If the file has been processed significantly more recently than the DAG has been updated, then its safe to assume that the DAG is missing and can be marked inactive.

(cherry picked from commit f309ea78f7d8b62383bc41eac217681a0916382b)

Commit:
Add documentation on specifying a DB schema. (#22347)

* Add documentation on specifying a DB schema.

From request - https://github.com/apache/airflow/issues/17374#issuecomment-1060019956

Co-authored-by: Nick Shook <nick.shook@apple.com>
(cherry picked from commit 3bc0da326e7963eab9154913d5cafbc1be1c1a67)

aspain committed:
adding `on_execute_callback` to callbacks docs (#22362)

* adding on_execute_callback

`on_execute_callback` is not listed but is an available callback to use. Could not find a docs page to link the way that the rest of them have.

* Update docs/apache-airflow/logging-monitoring/callbacks.rst

Co-authored-by: Josh Fell <48934154+josh-fell@users.noreply.github.com>

* Update docs/apache-airflow/logging-monitoring/callbacks.rst

Co-authored-by: Josh Fell <48934154+josh-fell@users.noreply.github.com>
Co-authored-by: eladkal <45845474+eladkal@users.noreply.github.com>
(cherry picked from commit 179c5b67e9c838f2f6df2b2b9a4451bfa09db42d)

tanelk committed:
Fix Tasks getting stuck in scheduled state (#19747)

The scheduler_job can get stuck in a state, where it is not able to queue new tasks. It will get out of this state on its own, but the time taken depends on the runtime of current tasks - this could be several hours or even days.

If the scheduler can't queue any tasks because of different concurrency limits (per pool, dag or task), then on next iterations of the scheduler loop it will try to queue the same tasks. Meanwhile there could be some scheduled tasks with lower priority_weight that could be queued, but they will remain waiting.

The proposed solution is to keep track of dag and task ids, that are concurrecy limited and then repeat the query with these dags and tasks filtered out.

Co-authored-by: Tanel Kiis <tanel.kiis@reach-u.com>
(cherry picked from commit cd68540ef19b36180fdd1ebe38435637586747d4)

eladkal commented on airflow/config_templates/config.yml:
```diff
@@ -1767,6 +1767,14 @@
+      description: |
+        How often (in seconds) to check for stale DAGs (DAGs which are no longer present in
+        the expected files) which should be deactivated.
+      version_added: 2.3.0
```
```suggestion
      version_added: 2.2.5
```

BTW is it a feature or a bug fix?
Maybe the PR that added it should not be targeted to 2.2.5?

ephraimbuddy replied:
It's a bug fix. it was previously targeted for 2.3.0. Thanks for this

ephraimbuddy committed:
Update version added for `deactivate_stale_dags_interval` config (#22478)

(cherry picked from commit 24a0d6a6ad71cdb922fec3df59acbc0e9f2da39c)

ephraimbuddy committed:
Bump version to 2.2.5

ephraimbuddy committed:
Add 2.2.5 to CHANGELOG.txt and UPDATING.md

potiuk committed:
Synchronize setup.py files with new providers

potiuk committed:
Remove memorystore, not importable example

potiuk committed:
Remove provider tests failing collection because lack of cherry-picks

potiuk committed:
Limit docs build to apache-airlfow + docker-stack for 2.2 branch

potiuk committed:
Add `dbt` spelling

potiuk committed:
Remove tests that are not really needed in 2-2 branch

potiuk committed:
Make v2-2-specific CI workflow

This is a change in the workflow specific to v2-2-branch that
brings build-ci-images and tests in the same workflow.

This can only be done because v2-2 branch no longer accepts PRs
from forks and workflows are only one with direct pushes from
commmitters. Then the workflows have write access to the packages
and can be simplified (and can also contain different set
of images than the main workflow).

potiuk committed:
Remove provider check from pre-commits

potiuk committed:
Remove auto-generated LICENSES-ui from pre-commit

potiuk committed:
Replace timedelta.max with year long timdelta in test_manager (#22527)

Timedelta.max used in tests is not realistic and in some
circumstances, when it is added to date, it might cause
date OverflowError. Using long (but not 999999999 days long)
timedelta solves the problem.

(cherry picked from commit 18da1217d7ae593ff33c681353b027fac9252523)

yenchenLiu committed:
Check and disallow a relative path for sqlite (#22530)

(cherry picked from commit 96e880d6ab33e05a0dc1d1e2b2178ed1053490af)

potiuk committed:
Make regular expressions deep-copyable in Python 3.6

potiuk committed:
Fixed Dask TLS tests to work on Python 3.6

potiuk committed:
Remove cache timeout limit
